src/debuginfo.cpp

```diff
@@ -774,9 +775,11 @@ static int lookup_pointer(DIContext *context, jl_frame_t **frames,
 
     int fromC = (*frames)[0].fromC;
     int n_frames = inlineInfo.getNumberOfFrames();
-    if (n_frames == 0)
+    if (n_frames == 0) {
+        JL_UNLOCK(&codegen_lock);
```

Comment:
This can turn cases that doesn't segfault right now to segfault (even when there's only one julia thread) so this can cause regression and is not just a partial fix.

The only way I can think of right now is to use nogc lock on unmanaged thread and normal lock on managed ones. I think this should be safe as long as no safepoint/infinite loop present in this critical region. You can check if the thread is managed by checking if `ptls` and `ptls->safe_point` are both non-NULL.

Depending on if we still care about the robustness of the non-threading build, ideally we also need to change all the lock functions on the codegen lock to the unconditional (lower case) versions.